compiler/src/dotty/tools/dotc/transform/SpecializeFunctions.scala

```diff
@@ -0,0 +1,139 @@
+package dotty.tools.dotc
+package transform
+
+import TreeTransforms.{ MiniPhaseTransform, TransformerInfo }
+import ast.Trees._, ast.tpd, core._
+import Contexts.Context, Types._, Decorators._, Symbols._, DenotTransformers._
+import SymDenotations._, Scopes._, StdNames._, NameOps._, Names._
+
+import scala.collection.mutable
+
+/** Specializes classes that inherit from `FunctionN` where there exists a
+ *  specialized form.
+ */
+class SpecializeFunctions extends MiniPhaseTransform with InfoTransformer {
+  import ast.tpd._
+  val phaseName = "specializeFunctions"
+  override def runsAfter = Set(classOf[ElimByName])
+
+  /** Transforms the type to include decls for specialized applys  */
+  def transformInfo(tp: Type, sym: Symbol)(implicit ctx: Context) = tp match {
+    case tp: ClassInfo if !sym.is(Flags.Package) && (tp.decls ne EmptyScope) =>
+      var newApplys = Map.empty[Name, Symbol]
+
+      tp.parents.foreach { parent =>
+        List(0, 1, 2).foreach { arity =>
+          val func = defn.FunctionClass(arity)
+          if (parent.derivesFrom(func)) {
+            val typeParams = tp.cls.typeRef.baseType(func).argInfos
+            val interface  = specInterface(typeParams)
+
+            if (interface.exists && tp.decls.lookup(nme.apply).exists) {
```

Comment:
Instead of creating a class symbol with `specInterface` and checking if it actually exists, you can use `isSpecializableFunction` like in https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/FunctionalInterfaces.scala#L39